python/neml2/pyzag/interface.py

```diff
@@ -0,0 +1,333 @@
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+from pyzag import nonlinear
+
+import neml2
+from neml2.tensors import Tensor
+
+import torch
+
+
+def assemble_vector(axis, tensors, warn_unused=True):
+    """Assemble a LabeledVector from a collection of tensors
+
+    Args:
+        axis (LabeledAxis): axis to use to setup LabeledVector
+        tensor (dict of torch.tensor): dictionary mapping names to tensors
+
+    Keyword Args:
+        warned_unused (bool, default True): throw a warning if there is a missing or extra variable in tensors
+    """
+    random_tensor = next(iter(tensors.values()))
+    batch_shape = random_tensor.shape[:-1]
+    device = random_tensor.device
+
+    vector = neml2.LabeledVector.zeros(batch_shape, [axis], device=device)
+
+    if warn_unused and set(tensors.keys()) != set(axis.variable_names()):
+        raise Warning(
+            "Tensor names in provided tensors dict do not match the variable names on the axis"
+        )
+
+    for name, value in tensors.items():
+        vector.base[name] = Tensor(value, len(batch_shape))
+
+    return vector
+
+
+class NEML2PyzagModel(nonlinear.NonlinearRecursiveFunction):
+    """Wraps a NEML2 model into a `nonlinear.NonlinearRecursiveFunction`
+
+    Args:
+        model (NEML2 model): the model to wrap
+
+    Keyword Args:
+        exclude_parameters (list of str): exclude these parameters from being wrapped as a pytorch parameter
+        state_axis (str): name of the state axis
+        forces_axis (str): name of the forces axis
+        residual_axis (str): name of the residual axis
+        old_prefix (str): prefix on the name of an axis to get the old values
+        neml2_sep_char (str): parameter seperator character used in NEML2
+        python_sep_char (str): seperator character to use in python used to name parameters
+    """
+
+    def __init__(
+        self,
+        model,
+        exclude_parameters=[],
+        state_axis="state",
+        forces_axis="forces",
+        residual_axis="residual",
+        old_prefix="old_",
+        neml2_sep_char=".",
+        python_sep_char="_",
+        *args,
+        **kwargs,
+    ):
+        super().__init__(*args, **kwargs)
+
+        self.model = model
+        self.lookback = (
+            1  # Hard coded because there aren't any other options in NEML2 right now
+        )
```

Comment:
I guess this doesn't prevent people from modifying the attribute. Since python doesn't have a direct way of "freezing" an attribute, I think it's best to assert lookback == 1 at critical locations.

Reply:
Saving this for later: https://github.com/applied-material-modeling/neml2/issues/180